flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/metrics/JobVertexWatermarksHandler.java

```diff
@@ -0,0 +1,99 @@
+import org.apache.flink.runtime.rest.handler.RestHandlerException;
+import org.apache.flink.runtime.rest.handler.job.AbstractJobVertexHandler;
+import org.apache.flink.runtime.rest.handler.legacy.ExecutionGraphCache;
+import org.apache.flink.runtime.rest.handler.legacy.metrics.MetricFetcher;
+import org.apache.flink.runtime.rest.handler.legacy.metrics.MetricStore;
+import org.apache.flink.runtime.rest.messages.EmptyRequestBody;
+import org.apache.flink.runtime.rest.messages.JobIDPathParameter;
+import org.apache.flink.runtime.rest.messages.JobVertexMessageParameters;
+import org.apache.flink.runtime.rest.messages.job.metrics.JobVertexWatermarksHeaders;
+import org.apache.flink.runtime.rest.messages.job.metrics.Metric;
+import org.apache.flink.runtime.rest.messages.job.metrics.MetricCollectionResponseBody;
+import org.apache.flink.runtime.webmonitor.RestfulGateway;
+import org.apache.flink.runtime.webmonitor.retriever.GatewayRetriever;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executor;
+
+
+/**
+ * Handler that returns the watermarks given a {@link JobID} and {@link JobVertexID}.
+ */
+public class JobVertexWatermarksHandler extends AbstractJobVertexHandler<MetricCollectionResponseBody, JobVertexMessageParameters> {
+
+	private final MetricFetcher metricFetcher;
+
+	public JobVertexWatermarksHandler(
+			GatewayRetriever<? extends RestfulGateway> leaderRetriever,
+			Time timeout,
+			Map<String, String> responseHeaders,
+			MetricFetcher metricFetcher,
+			ExecutionGraphCache executionGraphCache,
+			Executor executor) {
+		super(leaderRetriever,
+			timeout,
+			responseHeaders,
+			JobVertexWatermarksHeaders.INSTANCE,
+			executionGraphCache,
+			executor);
+		this.metricFetcher = metricFetcher;
+	}
+
+	@Override
+	protected MetricCollectionResponseBody handleRequest(
+			HandlerRequest<EmptyRequestBody, JobVertexMessageParameters> request,
+			AccessExecutionJobVertex jobVertex) throws RestHandlerException {
+
+		String jobID = request.getPathParameter(JobIDPathParameter.class).toString();
+		String taskID = jobVertex.getJobVertexId().toString();
+
+		metricFetcher.update();
+		MetricStore.TaskMetricStore taskMetricStore = metricFetcher.getMetricStore().getTaskMetricStore(jobID, taskID);
+		if (taskMetricStore == null) {
+			return new MetricCollectionResponseBody(Collections.emptyList());
+		}
+
+		AccessExecutionVertex[] taskVertices = jobVertex.getTaskVertices();
+		List<Metric> metrics = new ArrayList<>(taskVertices.length);
+
+		for (AccessExecutionVertex taskVertex : taskVertices) {
+			String id = taskVertex.getParallelSubtaskIndex() + "." + MetricNames.IO_CURRENT_INPUT_WATERMARK;
+			String watermarkValue = taskMetricStore.getMetric(id);
+			if (watermarkValue != null) {
+				metrics.add(new Metric(id, watermarkValue));
+			}
+		}
+
+		return new MetricCollectionResponseBody(metrics);
```

Comment:
I'm wondering whether this is a good representation. Having the subtask index encoded in the metric name is a bit redundant, particularly so when thinking about the future where additional watermark metrics may be listed (output watermark, watermarks for 2 input operators)

Ideally it would just be an array of `{ currentInputWatermark : XXX }` where the subtaks index is just the index in the array, but since we can't guarantee that metrics for all subtasks are available we could instead opt for `{ subtaskIndex: YYY, currentInputWatermark : XXX }`

Reply:
I think that is a good idea. I would defer this to a follow-up though. The PR fixes a bug and already improves the status quo. Next, we can think about optimizations.

Reply:
This follow-up would be an API breaking change and not just an optimization, and I'd like to avoid the risk of carrying this technical debt into future releases where we can no longer change it easily.

If we both agree that we want to have another representation in 1.10 then let's do it in one go.

Reply:
At this point I think it would be premature optimization. It is more important to fix this, then to save a few bytes for the responses. We already save a good portion of the transferred bytes due to the new zero byte header. I'd rather make a breaking change for the next release (if really necessary), instead of leaving this broken.

There is also another good argument for keeping this representation: The response format is aligned with the general metrics response which keeps a consistent and easy to understand metrics format.

Reply:
Let me clarify that my goal is not to "save a few bytes", but to use a format that is easier to extend in a sensible manner while at the same time being more user-friendly.

While staying consistent with existing code isn't a terrible argument (in fact I'm fine with merging this as is due to that) the reality is that this endpoint simply behaves different than the general metrics endpoint I wouldn't mind it having a separate representation.

Reply:
Agreed, a revised format would be nicer, but there is nothing inherently broken about the current one. I feel better about moving forward and fixing the issue, which was my original intend. I'm open to doing some follow-up work though. As we discovered, the metrics endpoints generally would benefit from a more flexible request/response format.